sw/device/lib/testing/dif/mock_dif_uart_test.cc

```diff
@@ -0,0 +1,228 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/testing/mock_mmio.h"
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+
+extern "C" {
+#include "sw/device/lib/dif/dif_uart.h"
+#include "uart_regs.h"  // Generated.
+}  // extern "C"
+
+namespace dif_uart_test {
+using ::mock_mmio::MmioTest;
+using ::testing::Test;
+
+class DifUartTestBasic : public Test, public MmioTest {
+ protected:
+  /**
+   * Sets up the default UART test configuration, the expectations are that the
+   * individual tests can further adjust the relevant values.
+   */
+  void SetUp() override {
+    // NCO is calculated as NCO = 2^20 * fbaud / fpclk, so the following values
+    // should result in NCO of 1.
+    dif_uart_config_pointer()->baudrate = 1;
+    dif_uart_config_pointer()->clk_freq_hz = 1048576;
+
+    // Initialise to defaults
+    dif_uart_config_pointer()->parity_enable = kDifUartDisable;
+    dif_uart_config_pointer()->parity = kDifUartParityEven;
+  };
+
+  // Return by reference to minimise impact of individual tests on each other.
+  dif_uart_t &dif_uart() { return dif_uart_; }
+  dif_uart_config_t &dif_uart_config() { return dif_uart_config_; }
+
+  // Return by pointer - to be used in overriden SetUp routine of derived
+  // classes.
+  dif_uart_t *dif_uart_pointer() { return &dif_uart_; }
+  dif_uart_config_t *dif_uart_config_pointer() { return &dif_uart_config_; }
+
+ private:
+  dif_uart_t dif_uart_;
+  dif_uart_config_t dif_uart_config_;
+};
+
+class DifUartTestUartInitialised : public DifUartTestBasic {
+ protected:
+  /**
+   * Uses the setup from the parent class, and initialises the UART state
+   * structure without calling the dif_uart_init() routine.
+   */
+  void SetUp() override {
+    DifUartTestBasic::SetUp();
+    dif_uart_pointer()->base_addr = dev().region();
+  };
+};
```

Comment:
I would avoid sticking this initialization in a test fixture. I think tests are easier to read if all of the calls to `dif_uart_` functions (if there are any, at all) happen in the test itself.

Reply:
Forgot to mention this, you can also call `_init()` and set the necessary expections instead of modifying the `base_addr` directly. I don't have a strong preference either way.